target_postgres/connector.py

```diff
@@ -277,6 +277,53 @@ def pick_individual_type(jsonschema_type: dict):
         if "object" in jsonschema_type["type"]:
             return JSONB()
         if "array" in jsonschema_type["type"]:
+            # Select between different kinds of `ARRAY` data types.
+            #
+            # This currently leverages an unspecified definition for the Singer SCHEMA,
+            # using the `additionalProperties` attribute to convey additional type
+            # information, agnostic of the target database.
+            #
+            # In this case, it is about telling different kinds of `ARRAY` types apart:
+            # Either it is a vanilla `ARRAY`, to be stored into a `jsonb[]` type, or,
+            # alternatively, it can be a "vector" kind `ARRAY` of floating point numbers,
+            # effectively what pgvector is storing in its `VECTOR` type.
+            #
+            # Still, `type: "vector"` is only a surrogate label here, because other
+            # database systems may use different types for implementing the same thing,
+            # and need to translate accordingly.
+            """
+            Schema override rule in `meltano.yml`:
+
+            type: "array"
+            items:
+              type: "number"
+            additionalProperties:
+              storage:
+                type: "vector"
+                dim: 4
+
+            Produced schema annotation in `catalog.json`:
+
+            {"type": "array",
+             "items": {"type": "number"},
+             "additionalProperties": {"storage": {"type": "vector", "dim": 4}}}
+            """
+            if (
+                "additionalProperties" in jsonschema_type
+                and "storage" in jsonschema_type["additionalProperties"]
+            ):
+                storage_properties = jsonschema_type["additionalProperties"]["storage"]
+                if (
+                    "type" in storage_properties
+                    and storage_properties["type"] == "vector"
+                ):
+                    # On PostgreSQL/pgvector, use the corresponding type definition
+                    # from its SQLAlchemy dialect.
+                    from pgvector.sqlalchemy import (
+                        Vector,  # type: ignore[import-untyped]
+                    )
+
+                    return Vector(storage_properties["dim"])
```

Comment:
After learning more details about the Singer specification, I've now used the `additionalProperties` schema slot to convey additional type information, agnostic of the target database.

It might not be what you had in mind, but it seems to works well. 🤷

Reply:
Hi Andreas!

Reading through the [JSON Schema Draft 7 spec](https://json-schema.org/draft-07/json-schema-validation#rfc.section.6.5.6) and thinking about our own use of `additionalProperties` [in the SDK](https://github.com/meltano/sdk/blob/299acc046990db9b2e3d68a9bb8a72eb73511857/singer_sdk/typing.py#L723-L727) I think I have a different understanding of its use cases:

<blockquote>
This keyword determines how child instances validate for objects, and does not directly validate the immediate instance itself.

Validation with "additionalProperties" applies only to the child values of instance names that do not match any names in "properties", and do not match any regular expression in "patternProperties".

For all such properties, validation succeeds if the child instance validates against the "additionalProperties" schema.
</blockquote>

So I think it's used to constraint the schema that extra fields not included in an object's `properties` mapping should have.

What do you think?

Reply:
Hi Edgar,

thanks for your response. I am absolutely sure I failed on respecting any specifications. Please bear with me that I've not made friends with the baseline specification too much yet, and apologies that I take up your time.

It is clear that I've abused the `additionalProperties` field, and I will be happy to wait for you unlocking corresponding other attributes how to convey additional type information, as you suggested at https://github.com/meltano/sdk/issues/2102.

My patch was merely thought to exercise a few other details which may be needed to make this fly, just on the level of the target, and to see if the details could be re-assembled to make it actually runnable [^1], also as a learning exercise.

With kind regards,
Andreas.

[^1]: https://github.com/singer-contrib/meltano-examples/tree/main/to-database

Reply:
Let me know if you we should better close this PR, and let the topic sit until corresponding improvements made it to the SDK.

I probably can't help there, as I believe it needs more intrinsic knowledge and discussions amongst you and your colleagues.